src/rebar_otp_app.erl

```diff
@@ -27,51 +27,63 @@
 %% Public API
 %% ===================================================================
 
-compile(Config, App) ->
+compile(State, App) ->
     %% If we get an .app.src file, it needs to be pre-processed and
     %% written out as a ebin/*.app file. That resulting file will then
     %% be validated as usual.
     Dir = ec_cnv:to_list(rebar_app_info:dir(App)),
-    {Config2, App1} = case rebar_app_info:app_file_src(App) of
+    {State2, App1} = case rebar_app_info:app_file_src(App) of
                           undefined ->
-                              {Config, App};
+                              {State, App};
                           AppFileSrc ->
-                              {Config1, File} = preprocess(Config, Dir, AppFileSrc),
-                              {Config1, rebar_app_info:app_file(App, File)}
+                              {State1, File} = preprocess(State, Dir, AppFileSrc),
+                              {State1, rebar_app_info:app_file(App, File)}
                       end,
 
     %% Load the app file and validate it.
     AppFile = rebar_app_info:app_file(App1),
-    case rebar_app_utils:load_app_file(Config2, AppFile) of
-        {ok, Config3, AppName, AppData} ->
+    case rebar_app_utils:load_app_file(State2, AppFile) of
+        {ok, State3, AppName, AppData} ->
             AppVsn = proplists:get_value(vsn, AppData),
-            validate_name(AppName, AppFile),
-            %% In general, the list of modules is an important thing to validate
-            %% for compliance with OTP guidelines and upgrade procedures.
-            %% However, some people prefer not to validate this list.
-            case rebar_state:get(Config3, validate_app_modules, true) of
-                true ->
-                    Modules = proplists:get_value(modules, AppData),
-                    {validate_modules(Dir, AppName, Modules), rebar_app_info:original_vsn(App1, AppVsn)};
-                false ->
-                    {ok, rebar_app_info:original_vsn(App1, AppVsn)}
+            case validate_name(AppName, AppFile) of
+                ok ->
+                    %% In general, the list of modules is an important thing to validate
+                    %% for compliance with OTP guidelines and upgrade procedures.
+                    %% However, some people prefer not to validate this list.
+                    case rebar_state:get(State3, validate_app_modules, true) of
+                        true ->
+                            case rebar_app_discover:validate_application_info(App1) of
+                                true ->
+                                    {ok, rebar_app_info:original_vsn(App1, AppVsn)};
+                                Error ->
+                                    Error
+                            end;
+                        false ->
+                            {ok, rebar_app_info:original_vsn(App1, AppVsn)}
+                    end;
+                Error ->
+                    Error
```

Comment:
Could we move that entire thing to an independent function? We should try to avoid having code blocks this huge, they become hard to maintain fairly fast.